modules/runner/src/integrationTest/java/org/apache/ignite/internal/sql/engine/ItSqlMultiStatementTxTest.java

```diff
@@ -0,0 +1,301 @@
+
+        assertThat(cursors, hasSize(2));
+
+        validateSingleResult(cursors.get(0));
+        validateSingleResult(cursors.get(1));
+
+        verifyFinishedTxCount(1);
+    }
+
+    @Test
+    void basicTxStatements() {
+        List<AsyncSqlCursor<List<Object>>> cursors = fetchAllCursors(runScript(
+                "START TRANSACTION;"
+                        + "INSERT INTO test VALUES(0);"
+                        + "INSERT INTO test VALUES(1);"
+                        + "COMMIT;"
+
+                        + "START TRANSACTION;"
+                        + "COMMIT;"
+
+                        + "START TRANSACTION;"
+                        + "INSERT INTO test VALUES(2);"
+                        + "COMMIT;"
+        ));
+
+        verifyFinishedTxCount(3);
+
+        assertQuery("select count(id) from test")
+                .returns(3L).check();
+
+        cursors.forEach(AsyncSqlCursor::closeAsync);
+    }
+
+    @Test
+    void readMoreRowsThanCanBePrefetchedReadOnlyTx() {
+        String query = "START TRANSACTION READ ONLY;"
+                + "SELECT 1;"
+                + "SELECT id FROM big;"
+                + "COMMIT;";
+
+        List<AsyncSqlCursor<List<Object>>> cursors = fetchAllCursors(runScript(query));
+        assertThat(cursors, hasSize(4));
+
+        BatchedResult<List<Object>> res = await(cursors.get(2).requestNextAsync(BIG_TABLE_ROWS_COUNT * 2));
+        assertNotNull(res);
+        assertThat(res.items(), hasSize(BIG_TABLE_ROWS_COUNT));
+
+        verifyFinishedTxCount(1);
+
+        cursors.forEach(AsyncSqlCursor::closeAsync);
+    }
+
+    @Test
+    void readMoreRowsThanCanBePrefetchedReadWriteTx() {
+        String query = "START TRANSACTION;"
+                + "UPDATE big SET salary=1 WHERE id=1;"
+                + "SELECT id FROM big;"
+                + "COMMIT;";
+
+        AsyncSqlCursor<List<Object>> cursor = runScript(query);
+
+        assertTrue(cursor.hasNextResult());
+
+        AsyncSqlCursor<List<Object>> updateCursor = await(cursor.nextResult());
+        assertNotNull(updateCursor);
+        validateSingleResult(updateCursor, 1L);
+        assertTrue(updateCursor.hasNextResult());
+
+        AsyncSqlCursor<List<Object>> selectCursor = await(updateCursor.nextResult());
+        assertNotNull(selectCursor);
+
+        BatchedResult<List<Object>> res = await(
+                selectCursor.requestNextAsync(BIG_TABLE_ROWS_COUNT / 2));
+
+        assertNotNull(res);
+        assertThat(res.items(), hasSize(BIG_TABLE_ROWS_COUNT / 2));
+        assertEquals(1, txManager().pending(), "Transaction must not finished until the cursor is closed.");
+        assertFalse(selectCursor.nextResult().isDone());
+
+        await(selectCursor.requestNextAsync(BIG_TABLE_ROWS_COUNT)); // Cursor must close implicitly.
+        verifyFinishedTxCount(1);
+    }
+
+    @Test
+    void openedScriptTransactionRollsBackImplicitly() {
+        {
+            fetchAllCursors(runScript("START TRANSACTION;"));
```

Comment:
We have test `commitWithoutOpenTransactionDoesNothing` that checks commit without tx.
We have test `emptyTransactionControlStatement` that checks `start; commit;` (validates cursor data, as you suggest here).
We have this test that verifies that the transaction will be rolled back in case of user input error. 
What kind of test you want to add?

Reply:
This test has been reworked a bit and updated.